프로젝트 매니저 [STEP 2] Brody, Rowan

Tediousday93 opened the pull request:
# STEP 2 PR
@havilog
안녕하세요 하비! Brody, Rowan 조 STEP 2 PR 보내드립니다 🫡
정말 오래걸렸네요...

compositionalLayout / MVVM에 대해 자세히 알아보는 시간이었습니다..!

이번 리뷰도 잘 부탁드립니다~~

---

# 고민한 점
### 🔸 의존성 주입
각 타입의 프로퍼티에 기본값을 할당하기 보다는 이니셜라이저를 통해 주입받을 수 있도록 했습니다.

예를 들어, DateFormatter의 경우 TaskCellViewModel에서 사용되기 때문에 기본값을 할당하면 여러개의 인스턴스가 생성됩니다.

하나의 인스턴스만 활용하기 위해 DateFormatter를 최상위 뷰컨트롤러에 만들어 사용되는 타입으로 전달할 수 있도록 하였습니다.

모든 프로퍼티가 Protocol타입으로 정의되어있지 않기 때문에 아직 완벽하게 Testable한 코드는 아니지만 protocol만 정의해주면 testable하게 될 것 같습니다.

</br>

### 🔸 요구사항에 맞는 레이아웃 설계
초기에는 하나의 컬렉션 뷰에서 컴포지셔널 레이아웃으로 세 개의 테이블뷰가 보이도록 만들었지만 이는 확장성이 떨어지고 레이아웃 잡는 코드가 복잡하다는 것을 직접 코드를 작성하면서 느꼈습니다.

그래서 하나의 컬렉션 뷰에서 관리하는 것이 아닌 3개의 컬렉션 뷰를 만드는 것이 좋다고 생각했습니다.
하지만 3개의 컬렉션 뷰는 화면에 task를 보여주는 공통의 역할을 하기에 한단계 상위의 뷰컨에서 하나의 컬렉션 뷰 타입을 갖는 여러개의 인스턴스를 갖도록 만들었습니다. 
```swift
// mainViewController.swift
self.addChild(TaskCollectionViewController(viewModel: todoViewModel,
                                           dateFormatter: dateFormatter))
self.addChild(TaskCollectionViewController(viewModel: doingViewModel,
                                           dateFormatter: dateFormatter))
self.addChild(TaskCollectionViewController(viewModel: doneViewModel,
                                           dateFormatter: dateFormatter))
```

그리고 추후 특정 뷰모델에만 기능이 추가되었을 경우. 예를 들어 `DoingViewModel`에만 특정 기능이 추가되었을 경우를 대비해서 하나의 뷰모델타입에서 `enum`타입을 이용해 분기처리하는 것이 아닌 컬렉션 뷰에 매칭되는 각각의 구체적인 뷰모델을 갖도록 만들었습니다. 

```swift
// 뷰모델에서 공통적으로 어떤 함수를 요청하고 받을 지를 추상화한 프로토콜
protocol TaskListViewModel { }

final class TodoViewModel: TaskListViewModel { }
final class DoingViewModel: TaskListViewModel { }
final class DoneViewModel: TaskListViewModel { }
````

</br>

### 🔸 TaskListViewModel - taskList 감지하기
`@Published` attribute를 설정한 프로퍼티가 publisher를 통해 값을 발행하는 타이밍이 프로퍼티 `willSet`이라는 것을 공식문서를 통해 알게 되었습니다.

이에 따라 $를 통해 해당 프로퍼티의 퍼블리셔를 이용하여 sanpshot을 dataSource에 apply하는 메서드를 바인딩하게 되면,
CollectionView에 taskList의 변경사항이 반영되지 않은 채로 snapshot apply가 진행되어 Out of range 크래시가 발생하는 문제가 있었습니다. (ex. createTask의 경우)

이를 해결하기 위해 taskList의 didSet과 PassthroughSubject를 활용하여 taskList가 set 되어진 이후 스트림에 변경된 list를 send 했습니다.

```swift
var taskList: [Task] = [] {
    didSet {
        currentTaskSubject.send((taskList, isUpdating))
    }
}
let currentTaskSubject = PassthroughSubject<([Task], Bool), Never>()
```

</br>

### 🔸 View 관련 Filetree
화면을 기준으로(main, detail, changeWorkState) 폴더를 나누고 매칭되는 뷰와 뷰모델을 모아서 정리했습니다.

</br>

# 조언을 얻고싶은 점

### 🔸 반응형 프레임워크 + Diffable DataSource에서 업데이트 시 전체가 reload되는 문제.
현재 `TodoViewModel`이라는 뷰모델 내의 `taskList`프로퍼티가 변경되었을 때 이를 구독하고 있는 `taskCollectionViewController`에서 변경된`taskList`을 받아, 이를 이용해 스냅샷을 업데이트하고 있습니다.

```swift
private func bindViewModelToView() {
    viewModel
        .currentTaskSubject
        .sink { taskList, isUpdating in
            isUpdating ? self.reloadDataSourceItems() : self.applyLatestSnapshot(taskList)

            self.viewModel.setState(isUpdating: false)
        }
        .store(in: &bindings)
}

private func applyLatestSnapshot(_ taskList: [Task]) {
    let taskIDList = taskList.map { $0.id }

    let section = Section.main(count: viewModel.taskList.count)
    var snapshot = Snapshot()
    snapshot.appendSections([section])
    snapshot.appendItems(taskIDList, toSection: section)
    dataSource?.apply(snapshot)
}

private func reloadDataSourceItems() {
    guard var snapshot = dataSource?.snapshot() else { return }
    let taskListID = viewModel.taskList.map { $0.id }

    snapshot.reloadItems(taskListID)
    dataSource?.apply(snapshot)
}
```

하지만 이 코드는 전달받은 `taskList`에 해당하는 `id`들을 전부 가져와 `reloadItems`를 하고 있기 때문에 특정 셀만 변경되었는데도 불구하고 전체 아이템이 reload되는 문제를 갖고 있습니다.

반응형 프레임워크를 사용하기 전에는 수정되는 task의 id를 반환해서 그 id만 reload한 뒤에 `dataSource.apply(snapshot)`을 호출해서 셀 하나만 업데이트 되도록 만들었는데 현재 상태에서 이와 같이 해결하는 방법을 찾지 못했습니다.

이 문제를 어떻게 해결하면 좋을까요?

</br>

### 🔸 ViewModel이 TaskList를 갖고 있어도 되는가
현재 MainViewModel과 TaskListViewModel에 프로퍼티로 taskList를 가지고 있습니다. 이에 따라 해당 list에 append / remove 하는 등의 로직을 view model이 가지게 되는데 이는 뷰 로직이 아니라 데이터 로직으로 생각될 수도 있을 것 같습니다.

이를 해결하기 위해 Service에서 fetch 할 때, Publisher<[Task], Never>를 반환해주는 식으로 바꿔보는게 어떨까 생각해봤습니다..

하비의 의견은 어떠신가요??? 🤔

</br>

### 🔸 cellProvider에서 셀을 dequeue했을 때 reuseIdentifier가 없을 때 처리
현재 컬렉션 뷰를 갖고있는 `TaskCollectionViewController`안에서 `DiffableDatasource`에 전달하는 `cellProvider`는 다음과 같이 구현되어있습니다.
```swift
private func cellProvider(_ collectionView: UICollectionView, indexPath: IndexPath, identifier: Task.ID) -> UICollectionViewCell? {
    guard let cell = collectionView.dequeueReusableCell(
        withReuseIdentifier: TaskCell.identifier,
        for: indexPath
    ) as? TaskCell else {
        fatalError("cell does not have a reuseIdentifier")
    }

    guard let task = self.viewModel.taskList.filter({ $0.id == identifier }).first else {
        fatalError("There is nonexistent identifier")
    }

    let taskCellViewModel = TaskCellViewModel(task: task, dateFormatter: dateFormatter)
    cell.provide(viewModel: taskCellViewModel)

    return cell
}
```

snapshot에서 주입하는 id가 뷰모델의 `taskList`와 매칭되지 않을 때 `fatalError`로 크래시나게 만들어주고 있습니다. 

**fatalError를 선택한 이유**
* guard 문에서 else에 걸렸을 때 `UICollectionViewCell()`을 리턴해준 경우, cell이 dequeueReusableCell을 통해 반환되지 않았다는 크래시가 발생함

반면에, UICollectionViewDataSource를 채택했을 때는 cellForItemAt 메서드에서 빈 UICollectionViewCell을 반환하는 것이 가능했습니다.

앱을 일부러 크래시 내는 코드는 피하는 것이 좋다고 생각하는데, DiffableDataSource에서도 fatalError를 사용하지 않고 빈 Cell을 반환해주는 방법을 활용할 수 있을까요?

seunghyunCheon committed:
feat: Todo모델 생성

Tediousday93 committed:
feat: ToDoTableViewController, Cell 생성

seunghyunCheon committed:
feat: 메인화면 투두스택뷰 추가

Tediousday93 committed:
feat: TodoListViewModel,  TodoViewModel 생성

seunghyunCheon committed:
feat: safe subscript추가

Tediousday93 committed:
feat: TodoTableViewController 이니셜라이저 정의

seunghyunCheon committed:
feat: 메인화면에 세개의 테이블뷰가 보이게 구성

Tediousday93 committed:
feat: HeaderView 정의

Tediousday93 committed:
docs: Update README

seunghyunCheon committed:
feat: DoingTableViewController, DoneTableViewController 생성

Tediousday93 committed:
feat: TodoTableViewCell contents label 정의

mainViewController layout safeArea로 변경

seunghyunCheon committed:
refactor: TodoCell UI수정

Tediousday93 committed:
remove: tableview controller 삭제

Tediousday93 committed:
refactor: mainViewController collectionView 정의

seunghyunCheon committed:
feat: DiffableDataSource, cell등록함수 구현

seunghyunCheon committed:
feat: Providable 프로토콜 구현

Tediousday93 committed:
feat: CollectionViewModel 정의

seunghyunCheon committed:
feat: 메인화면 DiffableDataSource 설정

Tediousday93 committed:
remove: main view protocol 폴더 삭제

Tediousday93 committed:
feat: collection view model update, add, remove 메서드 정의

seunghyunCheon committed:
feat: 메인화면 collectionView Datasource 설정

seunghyunCheon committed:
feat: Todo모델 Identifiable채택후 Hashable에서 ID로 변경

Tediousday93 committed:
refactor: collection view layout 가로 스크롤 configuration으로 변경

seunghyunCheon committed:
feat: layout에  HeaderView추가

Tediousday93 committed:
refactor: collection view compositional layout group edgeSpacing 적용

seunghyunCheon committed:
feat: NavBarItem, DetailViewController 구현

Tediousday93 committed:
feat: DetailView Textfeild shadow 적용

seunghyunCheon committed:
feat: textView layer설정

Tediousday93 committed:
rename: model명 Todo에서 Task로 변경

seunghyunCheon committed:
feat: UITextField Publisher 생성

Tediousday93 committed:
feat: UIControl+Combine 파일 생성, UIControlSubscription 클래스 정의

seunghyunCheon committed:
feat: textView Publisher적용 및 구독

Tediousday93 committed:
refactor: Combine UIControlPublisher UIControler Extension 내부로 위치 변경 및 내용 수정

seunghyunCheon committed:
feat: UIControl EventSubscription, EventPublisher 반환함수 정의

Tediousday93 committed:
feat: UITextField, UIDatePicker extension + Combine 추가

seunghyunCheon committed:
feat: isEditingDone상태에 따라 Done버튼 활성화

Tediousday93 committed:
refactor: detail view model date 프로퍼티 추가

Tediousday93 committed:
feat: TaskRepository 정의

seunghyunCheon committed:
feat: Task추가 기능 구현

Tediousday93 committed:
feat: DetailViewController updateTask 관련 기능 추가

nested enum Mode 추가
repository, viewmodel update 관련 메서드 추가

seunghyunCheon committed:
feat: 셀 업데이트 시 snapshot reloadItems하는 기능 구현, 메서드 위치 수정

Tediousday93 committed:
feat: Collection view cell 내부 swipe action을 위한 scrollview 추가

seunghyunCheon committed:
feat: 셀 Delete UI 설정

Tediousday93 committed:
Revert "feat: 셀 Delete UI 설정"

This reverts commit 92f47f3a97c1c6ce586615ddb9bde971dd816e81.

Tediousday93 committed:
Revert "feat: Collection view cell 내부 swipe action을 위한 scrollview 추가"

This reverts commit 44de5459d5934bef33868fc2b9a5732f64786f25.

seunghyunCheon committed:
refactor: 메인화면에서 세개의 컬렉션뷰 사용하도록 변경

Tediousday93 committed:
refactor: main view controller에 각 workState 별 ViewController addChild

seunghyunCheon committed:
feat: MainViewModel 구현

Tediousday93 committed:
feat: mainViewModel의 assignChildViewModel(of:) 메서드 정의

해당 메서드 사용을 위한 추상화로 TaskCollectionViewController, CollectionViewModel protocol 정의

seunghyunCheon committed:
feat: CollectionViewModel, TaskCollectionViewController 프로토콜 채택

Tediousday93 committed:
refactor: CollectionViewModel 프로토콜 associatedtype 추가

associatedtype을 추가함에 따라, CollectionViewModel 타입으로 활용 시 any 키워드 추가

seunghyunCheon committed:
refactor: mainViewModel에서 TodoCollectionViewModel에 데이터주입 및 snapshot적용

Tediousday93 committed:
refactor: DoingCollectionViewModel, DoingViewController 추상화에 맞게 수정

seunghyunCheon committed:
refactor: DoneCollectionViewModel, DoneViewController 추상화에 맞게 수정

Tediousday93 committed:
refactor: MainViewModel, TodoCollectionViewModel 인터페이스 수정

fetchTaskList -> configureCollectionViewModels
TodoCollectionViewModel 접근제어 수정

seunghyunCheon committed:
refactor: doingViewModel, doneViewModel snapshot로직 수정

seunghyunCheon committed:
refactor: task update하는 부분 메인뷰컨의 책임으로 위임

Tediousday93 committed:
refactor: TaskFetchDelegate 객체 MainViewModel로 변경

TaskFetchDelegate -> DetailViewControllerDelegate로 이름 변경

seunghyunCheon committed:
feat: Detail화면 생성, 수정에 따라 UI상호작용 여부 설정

Tediousday93 committed:
feat: cell trailingSwipeAction 추가 (Delete)

Tediousday93 committed:
refactor: MainTaskService, CollectionTaskService 분리 중

seunghyunCheon committed:
refactor: MainTaskService, CollectionTaskService분리 완료

seunghyunCheon committed:
feat: TodoCollectionViewModel remove 기능 구현

Tediousday93 committed:
refactor: CollectionViewModel 추상화 중도 커밋

TaskCollectionViewModel 정의

Tediousday93 committed:
refactor:  이전 커밋 누락된 변경사항 추가

seunghyunCheon committed:
remove: 보일러플레이트 코드 삭제

seunghyunCheon committed:
reafactor: TaskCollectionViewController 생성 후 모든 뷰컨트롤러 대체

seunghyunCheon committed:
docs: 리드미 수정

Tediousday93 committed:
docs: 리드미 불필요한 부분 삭제

Tediousday93 committed:
chore: 파일트리 수정

seunghyunCheon committed:
refactor: TaskListViewModel 프로토콜  정의

TaskCollectionViewController에서 컬렉션뷰와 DataSource갖도록 변경.
TaskListViewModel 구체타입 TodoViewModel 정의.
TaskListViewModel 프로토콜 기본구현

Tediousday93 committed:
refactor: TaskCollectionViewController 수정, Doing, Done ViewModel 정의

TaskCollectionViewController dataSource 프로퍼티, updateDataSource 메서드 정의

seunghyunCheon committed:
refactor: MainViewModel에서 전체 데이터를 관리하도록 변경

MainTaskService -> TaskStorageService로 변경

Tediousday93 committed:
refactor: MainViewController의 컴파일 오류 해결

TaskListViewModel 구체 타입 taskWorkState 초기값을 갖도록 수정

seunghyunCheon committed:
fix: 메인화면에서 데이터 패치되지않는 문제를 바인딩을 통해 해결

Tediousday93 committed:
refactor: DetailViewModel delegate 수정, collection view dataSource apply 오류 수정 중

seunghyunCheon committed:
fix: CollectionView DataSource apply오류 수정

@Published에서 CurrentValueSubject로 변경해서 Task가 생성되게 수정
TaskList에 didSet 추가

Tediousday93 committed:
refactor: DetailViewModelDelegate에 setState(isUpdating:) 메서드 추가

추가된 메서드를 통해 task create, update 작업을 구분할 수 있도록 수정
TaskListViewModel에 작업 분기를 나눠줄 프로퍼티 isUpdating 추가

seunghyunCheon committed:
refactor: DetailViewController에 navigationController Embed.

기존에 navigationBar를 넣어주던 방식을 제거하고 navigationController를 삽입.
modal present방식 formsheet로 변경.

Tediousday93 committed:
refactor: HeaderView viewmodel 추가 및 binding

seunghyunCheon committed:
refactor: CellView에 CellViewModel 프로퍼티 추가 및 binding

HeaderViewCell 접근제어자 수정.

Tediousday93 committed:
remove: 불필요한 파일 삭제

Tediousday93 committed:
refactor: DetailViewModel Mode 열거형 ViewModel로 이동, HeaderView Height 설정

seunghyunCheon committed:
feat: setupLongTapGestureRecognizer, handleLongPress함수 구현

셀을 길게 눌렀을 때 선택된 셀에서만 이벤트가 실행되도록 설정.

Tediousday93 committed:
feat: MovePopoverViewController 정의

seunghyunCheon committed:
feat: ChangeWorkStateViewModel 구현

ChangeWorkStateViewController 뷰모델과 바인딩하는 함수 구현.
셀 길게 누를 시 popoverPresentationController띄워지는 기능 구현.

Tediousday93 committed:
fix: section header가 업데이트 되지 않는 문제 해결

create, update, delete task cell 진행 시 , 모두 snapshot reloadSections 호출하도록 변경

seunghyunCheon committed:
feat: popover된 뷰컨에서 버튼 클릭 시 Task 상태가 변하도록 변경.

Task 상태를 변경하는 프로토콜 ChangeWorkStateViewModelDelegate 정의

Tediousday93 committed:
feat: popover button을 통해 workState를 변경하는 기능 구현

Tediousday93 committed:
fix: array safe subscript를 통한 값 할당이 되지 않던 문제 수정

seunghyunCheon committed:
fix: 자식 뷰 컨트롤러의 뷰모델의 델리게이트를 메인 뷰모델로 설정

Tediousday93 committed:
refactor: MainViewController, MainViewModel, TaskStorageService 의존성 주입

seunghyunCheon committed:
refactor: Section에 연관값을 가져 섹션헤더에서 변화를 감지할 수 있도록 변경.

TaskListViewModel에서 Section연관값, sectionInfo 프로퍼티 설정.
TaskCollectionViewController에서 제네릭 Section을 받도록 설정.
MainViewModel UIKit 삭제 후 뷰컨트롤러가 아닌 뷰모델을 받도록 변경.

Tediousday93 committed:
fix: 특정 상황에서 delete할 때 outOfRange가 발생하는 오류 수정

update 시 TaskListViewModel의 isUpdating이 true로 바뀐 것이 유지되고 delete 시 isUpdating을 따로 설정하지 않았음
따라서 delete 할 때 isUpdating이 true이므로 sanpshot의 reload메서드가 호출되어 out of range가 발생.

해당 오류는 viewModel의 delete 메서드에서 isUpdating을 false로 설정하도록 변경하여 해결

seunghyunCheon committed:
fix: 셀 수정 후 이동시 남아있는 문제 setState함수호출를 통해  수정

Tediousday93 committed:
refactor: NavigationBar border, stackView background 설정

seunghyunCheon committed:
feat: CALayer 익스텐션 정의, 스택뷰의 아래에만 border 적용

seunghyunCheon committed:
feat: 셀 날짜 만료되었을 때 빨간색으로 변경하는 로직 구현

Tediousday93 committed:
refactor: HeaderView 하단 padding 추가

seunghyunCheon committed:
refactor: TaskListViewModel에 있던 Section을 TaskCollectionViewController로 이동

Tediousday93 committed:
refactor: 코드 컨벤션, 어색한 네이밍 수정 및 불필요한 코드 삭제

seunghyunCheon committed:
refactor: TaskCollectionViewContrller에서 collectionViewDelegate, Gesture관련 기능 extension으로 이동

seunghyunCheon committed:
refactor: ReusableIdentifier 프로토콜 정의

seunghyunCheon committed:
refactor: dequeue되지않은 셀을 반환할 시 fatalerror뿜도록 변경

seunghyunCheon committed:
refactor: cellProvider내부의 identifier가 존재하지 않을 경우 fatalError 메세지 수정

havilog commented on ProjectManager/ProjectManager/SceneDelegate.swift:
```diff
@@ -7,15 +7,32 @@
+            Task(title: "hi", date: Date(), body: "body", workState: .todo),
+            Task(title: "hi", date: Date(), body: "body", workState: .todo),
+            Task(title: "hi", date: Date(), body: "body", workState: .todo)
+        ]
```
이 default로 넣어주는 아이들은 요구사항인가요??
테스트용으로 보여서요

seunghyunCheon replied:
요구사항이 아니라서 지웠습니다!

havilog commented on ProjectManager/ProjectManager/Model/Task.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+struct Task: Identifiable {
```
`Task`라는 이름은 좀 알아보기 힘든거 같아요.
FirstParty에서 async/await에서 쓰이는 Task와 같은 단어라서
다른 네이밍을 찾아보는게 나을거 같다고 생각됩니다!

seunghyunCheon replied:
`Plan`으로 변경했습니다!

havilog commented on ProjectManager/ProjectManager/Model/Task.swift:
```diff
@@ -0,0 +1,24 @@
+    let title: String
+    let date: Date
+    let body: String
+    var workState: WorkState
```
workState가 바뀐다면, id도 바껴서 새 객체를 만들어주는거랑 비슷해보이는데, variable로 선언하신 이유가 있을까요?

seunghyunCheon replied:
`workState`가 바뀌면 `id`도 바껴서 새 객체를 만들어주는 것과 비슷하다는 말씀이 이해가 잘 되지 않습니다! 
저희는 `todo`섹션의 셀이 `doing`섹션으로 이동할 때 `id`는 동일하고 `workState`만 변경이 되어야 `DiffableDataSource`에서 변경사항에 대한 스냅샷 차이를 계산할 것이라고 생각이 들어서 `var`로 설정해주었습니다.

havilog commented on ProjectManager/ProjectManager/Model/Task.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+struct Task: Identifiable {
```
테스트 짜셨는지 확인 안해봤지만,
테스트를 위해서 보통 모델들은 기본적으로 equatable 하면 좋아요

Tediousday93 replied:
Equatable 채택했습니다!

havilog commented on ProjectManager/ProjectManager/Model/WorkState.swift:
```diff
@@ -0,0 +1,34 @@
+    case doing
+    case done
+    
+    var text: String {
+        switch self {
+        case .todo:
+            return "TODO"
+        case .doing:
+            return "DOING"
+        case .done:
+            return "DONE"
+        }
+    }
```
이건 그냥 권장도 아니고 개인 의견인데,
WorkState자체에 rawValue로 주거나, `CustomStringConvertible`을 채택시켜주는 방법도 고려해 볼 수 있을거 같아요

Tediousday93 replied:
`case.rawValue` 보다는 `case.text`가 좀 더 보기 편하다고 생각해서 현재 상태를 유지했습니다!

CustomStringConvertible 같은 경우, 공식문서 상에서 description 프로퍼티를 정의한 후 해당 프로퍼티에 직접 접근하는 것은 권장되지 않고 `String(describing:)`메서드를 호출해야하기 때문에 이 부분도 `.text`가 더 한 눈에 들어올 것 같다고 생각합니다. 🤔

havilog commented on ProjectManager/ProjectManager/Protocols/TaskListViewModelDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+protocol TaskListViewModelDelegate {
+    func createTask(_ task: Task)
+    func updateTask(_ task: Task)
+    func deleteTask(id: UUID)
```
개인적인 취향이니까 참고만 해주세요 ㅎㅎ

```suggestion
    func create(task: Task)
    func update(task: Task)
    func deleteTask(id: UUID)
```

Tediousday93 replied:
하비의 suggestion으로 변경했습니다~!  👍

havilog commented on ProjectManager/ProjectManager/Protocols/TaskListViewModel.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    func createTask(_ task: Task)
+    func updateTask(_ task: Task)
+    func deleteTask(at index: Int)
```
TaskListViewModelDelegate에서는 id 로 지우고,
여기에서는 index로 지우는데
어떤 기준일까요?

seunghyunCheon replied:
컬렉션 뷰 내에서는 index가 존재하기 때문에 한 번에 index에 접근해서 지웠고, `service`내부의 plan배열에는 workState별로 나눠져있지 않아서 인덱스로 접근하기가 까다로워 id를 통해 해당 요소의 인덱스를 구했습니다. 

추후 todo, doing, done에 해당하는 배열 3개를 가지면서 이 부분을 해결할 것 같습니다.

havilog commented on ProjectManager/ProjectManager/Protocols/DetailViewModelDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+protocol DetailViewModelDelegate: AnyObject {
+    func createTask(_ task: Task)
+    func updateTask(_ task: Task)
```
이 두 함수들은 계속 반복적으로 존재하는데,
TaskHandlable(가칭)과 같은 프로토콜 하나로 뺄 수는 없었을까요?

Tediousday93 replied:
`PlanManagable` 이라는 프로토콜로 create(task:), update(task:) 메서드를 뽑아 줬습니다!

havilog commented on ProjectManager/ProjectManager/Extensions/Combine/UITextField+Combine.swift:
```diff
@@ -0,0 +1,17 @@
+            .map { self.text ?? "" }
+            .eraseToAnyPublisher()
+    }
+}
```
저도 옛날에는 이런 Extension들 많이 만들어 썼었는데요,
이게 쓰면 코드가 이뻐지긴 하는데,
다른 협업하는 사람들이 이코드를 모르고 안쓰면 레거시가 되는 코드라서
이런 간단한 코드는 그냥 뷰에 두는게 나은거 같아요.

또한 여기서 self.text ?? "" 이 부분은 비즈니스로직이라고 불릴 수도 있을거 같은데,
이런 부분은 디버깅 하기도 힘들어지구요

Tediousday93 replied:
닐 병합 연산자를 삭제하고 AnyPublisher<String?, Never>로 수정하였습니다!

havilog commented on ProjectManager/ProjectManager/Protocols/ChangeWorkStateViewModelDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+//
+//  ChangeWorkStateViewModelDelegate.swift
+//  ProjectManager
+//
+//  Created by Brody, Rowan on 2023/05/29.
+//
+
+import Foundation
+
+protocol ChangeWorkStateViewModelDelegate: AnyObject {
+    func changeWorkState(taskID: UUID, with workState: WorkState)
+}
```
보통 프로토콜은 이렇게 프로토콜 폴더에 안모아두고,
해당 프로토콜을 준수하는 객체나,
연관있는 네임스페이스에 정의하는게 나은거 같아요.
아니면 이 프로토콜이 어디서 뭐하는 역할인지 확인하려고 페이지 점핑이 일어날 수 있어서요

Tediousday93 replied:
프로토콜을 적절한 위치에 옮겨주었습니다 ~~

`ReusableIdentifier` 프로토콜의 경우에는 여러 타입에서 사용할 수 있을 것 같아서 따로 빼두었는데
이 프로토콜도 Cell 위쪽으로 옮기는게 나을까요? 🤔

havilog replied:
요건 그냥 Extension에 넣으면 될 거 같아요

havilog commented on ProjectManager/ProjectManager/ChangeWorkStateView/ChangeWorkStateViewController.swift:
```diff
@@ -0,0 +1,110 @@
+    }()
+    
+    private let viewModel: ChangeWorkStateViewModel
+    private var bindings = Set<AnyCancellable>()
```
요것도 취향이긴 한데, 보통 AnyCancellable를 담는 변수명을 cancellables로 하는게 일반적이라고 알고있어요

Tediousday93 replied:
bindings를 모두 cancellables로 수정했습니다!

havilog commented on ProjectManager/ProjectManager/ChangeWorkStateView/ChangeWorkStateViewController.swift:
```diff
@@ -0,0 +1,110 @@
+        
+        let safe = view.safeAreaLayoutGuide
+        
+        NSLayoutConstraint.activate([
+            stackView.leadingAnchor.constraint(equalTo: safe.leadingAnchor, constant: 5),
+            stackView.trailingAnchor.constraint(equalTo: safe.trailingAnchor, constant: -5),
+            stackView.topAnchor.constraint(equalTo: safe.topAnchor, constant: 5),
+            stackView.bottomAnchor.constraint(equalTo: safe.bottomAnchor, constant: -5)
```
요것도 취향차이긴 한데,
굳이 지역변수 만들어서 메모리 안잡아먹고 view.safeAreaLayoutGuide로 풀어쓰는게 더 가독성이 좋다는 사람들도 있어서 참고차 말씀드려여
고칠 정도까진 아닌듯

havilog commented on ProjectManager/ProjectManager/ChangeWorkStateView/ChangeWorkStateViewController.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    private func bindViewModelToView() {
+        viewModel.$firstText
+            .sink { firstText in
+                self.firstButton.setTitle(firstText, for: .normal)
+            }
+            .store(in: &bindings)
+        
+        viewModel.$secondText
+            .sink { secondText in
+                self.secondButton.setTitle(secondText, for: .normal)
+            }
+            .store(in: &bindings)
```
leak 나겠네여 ㅎㅎ

seunghyunCheon replied:
`weak self`붙여주었습니다!

havilog commented on ProjectManager/ProjectManager/ChangeWorkStateView/ChangeWorkStateViewController.swift:
```diff
@@ -0,0 +1,110 @@
+        self.dismiss(animated: true)
+    }
+    
+    private func bindViewModelToView() {
```
함수명 bind()로만 해도 괜찮을거 같아여 ㅎㅎ

seunghyunCheon replied:
`DetailViewController`의 경우 뷰에서 뷰모델로 바인드하는 부분과 뷰모델에서 뷰에 바인드하는 부분 두 가지가 존재해서 메서드 이름을 구분시켜주었습니다. 이러한 경우에도 `bind`안에 양방향 바인딩하는 부분을 넣어도 될까요?

seunghyunCheon replied:
해당 이름으로 수정하였습니다!

havilog commented on ProjectManager/ProjectManager/DetailView/DetailViewController.swift:
```diff
@@ -0,0 +1,209 @@
+        bodyTextView.isUserInteractionEnabled = isEnable
+    }
+    
+    private func bindViewToViewModel() {
+        assign(publisher: titleTextfield.textPublisher, keyPath: \.title)
+        assign(publisher: bodyTextView.textPublisher, keyPath: \.body)
+        assign(publisher: datePicker.datePublisher, keyPath: \.date)
+    }
```
여기두 leak 날거 같아여

Tediousday93 replied:
ViewModel에 Input/Output을 정의한 뒤 Event Stream이 연결되도록 설계하면서
assign operator는 사용하지 않게 되었습니다. 👀

havilog commented on ProjectManager/ProjectManager/DetailView/DetailViewController.swift:
```diff
@@ -0,0 +1,209 @@
+        bodyTextView.isUserInteractionEnabled = isEnable
+    }
+    
+    private func bindViewToViewModel() {
```
네이밍을 bindAction, bindState로 나눠도 괜찮겠네여

seunghyunCheon replied:
해당 이름으로 수정하였습니다!

havilog commented on ProjectManager/ProjectManager/DetailView/DetailViewController.swift:
```diff
@@ -0,0 +1,209 @@
+        viewModel
+            .isEditingDone
+            .sink { isEditingDone in
+                let rightBarButtonItem = self.navigationItem.rightBarButtonItem
```
지역변수로 잡아주는 이유가 뭔가여??
+여기도 weak

havilog commented on ProjectManager/ProjectManager/DetailView/DetailViewModel.swift:
```diff
@@ -0,0 +1,69 @@
+import Foundation
+import Combine
+
+final class DetailViewModel {
```
어떤 함수, 변수가 뷰컨에서 접근이 가능하고, 불가능한지
즉 viewmodel에 들어갈 input
뷰에서 옵저빙할 output이 명시되면
더 가독성이 좋은 코드가 될거 같아요.

seunghyunCheon replied:
중첩타입으로 viewModel에 들어갈 이벤트들을 `input`으로, 옵저빙할 요소로 `output`으로 명시해주었습니다.

havilog commented on ProjectManager/ProjectManager/MainView/MainViewController.swift:
```diff
@@ -0,0 +1,123 @@
+        return stackView
+    }()
+    
+    let mainViewModel: MainViewModel
```
UI Property 사이에 뷰모델이 ..?!

Tediousday93 replied:
DateFormatter는 UI Property가 아니라고 생각해서
UI property - internal property - private property 순으로 나열했습니당,,,

havilog commented on ProjectManager/ProjectManager/MainView/MainViewModel.swift:
```diff
@@ -0,0 +1,63 @@
+
+extension MainViewModel: ChangeWorkStateViewModelDelegate {
+    func changeWorkState(taskID: UUID, with workState: WorkState) {
+        service.changeWorkState(taskID: taskID, with: workState)
+        fetchTaskList()
+        distributeTask()
```
workState가 바뀔 때 마다 모든 TaskList를 fetch해와서 전체 리로드가 발생하는거 같아요.
메인 뷰모델에서 todoList, DoingList, DoneList를 가지고 있고,
CRUD에 성공했을 때 메모리에 들고 있는 배열의 element를 바꿔주는 방식은 혹시 고려해보셨을까요??

+ CRUD를 할 때 실패할 수도 있을거 같은데 지금 코드는 로컬 스토리지에서 실패하는 경우를 가정하고 있지 않아요.
이 부분도 고려해서 수정해주세여

Tediousday93 replied:
CRUD에 실패할 경우를 고려하기 위해 스트림을 만들 때 tryMap을 활용해 적절한 Error를 던질 수 있도록 해봤습니다!

말씀해주신 대로 CRUD에 성공하면 메모리에 들고 있는 배열의 plan 인스턴스를 바꿔주도록 변경했습니다. 👍

havilog commented on ProjectManager/ProjectManager/MainView/TaskCell/TaskCell.swift:
```diff
@@ -0,0 +1,106 @@
+        bindViewModelToView()
+    }
+    
+    private func bindViewModelToView() {
+        viewModel?.$title
+            .sink { title in
+                self.titleLabel.text = title
+            }
+            .store(in: &bindings)
+        
+        viewModel?.$body
+            .sink { body in
+                self.bodyLabel.text = body
+            }
+            .store(in: &bindings)
+        
+        viewModel?.$date
+            .sink { date in
+                self.dateLabel.text = date
+            }
+            .store(in: &bindings)
+        
+        viewModel?.$isDateExpired
+            .sink { state in
+                guard let state else { return }
+                
+                self.dateLabel.textColor = state ? .systemRed : .black
+            }
+            .store(in: &bindings)
+    }
```
셀에도 뷰모델이 필요한가?
에 대해서는 정답은 없긴 한데요,

제 생각을 말씀드리면, 뭔가 셀 자체적으로 API를 찌르거나 CRUD를 하거나,
캐싱같은 비지니스 로직들이 들어가게 된다면 필요하다고 보는데,

지금 같은 경우에는 데이터 조각 들(String)만 가지고 셀을 그릴 수 있을거 같아요

역할 분리와 오버엔지니어링은 한끗차이라서 요 부분도 잘 고려해보고 개발하시면 도움될거 같아요 ㅎㅎ

틀렸다는거 아닙니다!

Tediousday93 replied:
셀 뷰모델을 만들 때 Combine 활용에 대해서 계속 알아보고 있었습니다...
그래서 사용방법도 익힐 겸 바인딩을 해봤는데 하비 말씀대로 string을 그냥 텍스트에 넣어주기만 해도 괜찮았네요 !! 하하하하 😅

havilog commented on ProjectManager/ProjectManager/MainView/TaskCollectionView/DoingViewModel.swift:
```diff
@@ -0,0 +1,27 @@
+import Combine
+
+final class DoingViewModel: TaskListViewModel {
+    var taskList: [Task] = [] {
+        didSet {
+            currentTaskSubject.send((taskList, isUpdating))
+        }
+    }
```
사실 didSet이 편하긴 한데 ,,,
didSet은 테스트 하기 힘들어져서 저는 잘 안쓰긴 해요 ㅋㅋ

Tediousday93 replied:
Input, Output을 만들면서 삭제됐습니다~~

havilog reviewed (changes requested):
![image](https://github.com/yagom-academy/ios-project-manager/assets/57659933/d3eb6e38-ca44-49fa-8a41-31c054016586)

edit 눌렀을 때 timeout이 일어나고 있어요

> 🔸 반응형 프레임워크 + Diffable DataSource에서 업데이트 시 전체가 reload되는 문제.

코멘트 참고

> 🔸 ViewModel이 TaskList를 갖고 있어도 되는가

저는 개인적으로 `Publisher`는 가독성이 좋지 않고, 읽기 힘들다고 생각해서,
만들어도 큰 상관은 없을거 같긴 한데, 여러 방식들을 비교해보고 만들면 좋을거 같아요

> 🔸 cellProvider에서 셀을 dequeue했을 때 reuseIdentifier가 없을 때 처리

크래시는 진짜 최악의 상황에서만 발생해야하는거라..
reuseIdentifier가 없어서 셀 하나가 안보이는 짜친 이슈같은 경우에는,,
디버그일 때만 assert나 preconditionFailure같은거 찍어놓고
그냥 빈 셀 리턴시켜줘도 괜찮을거 같아요


프로토콜을 더 잘 쓰면 좀 더 제너럴하게 쓰이는 MVVM이 될 거 같아요!
두 분 모두 익숙하지 않은 프레임워크 사용하시느라 고생 많으셨구
조금 더 달려봅시다 : )

Tediousday93 committed:
refactor: DetailViewModel isEditingDone 퍼블리셔에 Error 추가, EditError 정의

body 프로퍼티 타입을 String? 으로 변경함에 따른 버튼 활성화 스트림 수정

seunghyunCheon committed:
refactor: textField Publisher에서 AnyPublisher<String?, Never>를 반환하도록 변경

seunghyunCheon committed:
refactor: Task -> Plan으로 이름변경

seunghyunCheon committed:
refactor: plan이름 사용하는 곳 카멜케이스 적용

Tediousday93 committed:
refactor: Plan 모델 Equatable 채택, Protocol에 선언된 메서드명 변경

createPlan(_ plan:) 형식의 메서드를 create(plan:)으로 변경
Model의 Test 용이성을 위한 Equatable 프로토콜 채택

seunghyunCheon committed:
refactor: plan포함한 변수명 등 카멜케이스로 수정, 메서드명 수정

Tediousday93 committed:
refactor: Protocol 상속을 통해 공통된 인터페이스 분리

PlanManagable 프로토콜에 create, update 인터페이스 선언
각 프로토콜을 연관된 타입 파일에 이동

seunghyunCheon committed:
refactor: bindings -> cancellables로 변수명 수정, PlanManagable위치 PlanStorageService로 이동

Tediousday93 committed:
refactor: DetailViewController assign 메서드 메모리 누수 방지를 위해 sink로 변경

Tediousday93 committed:
refactor: DetailViewController bind 메서드 이름 변경

bindViewToViewModel -> bindAction
bindViewModelToView -> bindState

seunghyunCheon committed:
refactor: mainViewModel에서 필터링된 list배열 프로퍼티를 갖도록 변경

Tediousday93 committed:
refactor: PlanListViewModel 설계 수정 및 구체타입 TodoViewModel 수정

seunghyunCheon committed:
refactor: doingViewModel, doneViewModel 프로퍼티 요구사항 변경, DetailViewModelDelegate 프로토콜 setState메서드 제거.

Tediousday93 committed:
refactor: DetailViewModel @Published 프로퍼티 래퍼 삭제, ChangeWorkStateViewModel 내부 프로퍼티 설계에 맞게 수정

seunghyunCheon committed:
refactor: 컬렉션뷰 스냅샷의 아이템 여러개가 리로드되는 상황 개선.

Tediousday93 committed:
refactor: DetailViewController, DetailViewModel 수정

DetailViewModel nested type Input, Output 정의
DetailViewModel cancellables 프로퍼티 추가
DetailViewController bind 메서드 수정

Tediousday93 committed:
refactor: DetailViewModel transform 메서드 수정

ViewController에서만 구독이 일어나도록 수정

seunghyunCheon committed:
refactor: 섹션의 연관값에 의해 전체 아이템이 리로드되는 문제 HeaderViewModel 바인딩을 통해 해결

Tediousday93 committed:
refactor: 오타수정 및 ChangeWorkStateViewModel Input, Output nestedtype 정의 중

seunghyunCheon committed:
refactor: Plan이동 시 전체가 리로드 되는 상황 개선.

ChangeWorkStateViewModel transform함수 정의. dismiss하는 output 반환.
MainViewModel changeWorkState함수 구현 변경.
PlanListViewModel protocol 이벤트 발생 시 전달되는 타입 변경.

seunghyunCheon committed:
refactor: MainViewModel 3개의 [Plan] 프로퍼티에서 하나의 프로퍼티로 변경

Tediousday93 committed:
refactor: leftBarButtonItem selector 메서드 이름 변경

Tediousday93 committed:
rename: 폴더 명 task -> plan으로 수정

seunghyunCheon committed:
refactor: publisher merge전 eraseToAnyPublisher함수 제거, TaskCell -> PlanCell로 폴더명 변경

seunghyunCheon committed:
docs: README 수정

seunghyunCheon committed:
docs: 리드미 수정

seunghyunCheon committed:
refactor: 테스트 데이터 제거

seunghyunCheon committed:
feat: 생성, 수정, 삭제 시 에러처리
